Index: linux-3.10.14/arch/mips/ralink/init.c
===================================================================
--- linux-3.10.14.orig/arch/mips/ralink/init.c
+++ linux-3.10.14/arch/mips/ralink/init.c
@@ -543,7 +543,7 @@ void prom_init_sysclk(void)
 #elif defined (CONFIG_RALINK_MT7628)
 	case 0:
 		reg = (*(volatile u32 *)(RALINK_SYSCTL_BASE + 0x10));
-		if (reg & 0x80)
+		if (reg & 0x40)
 		{
 			/* 40MHz Xtal */
 			mips_cpu_feq = 580 * 1000 * 1000;
