Index: linux-3.10.14/net/nat/hw_nat/ra_nat.c
===================================================================
--- linux-3.10.14.orig/net/nat/hw_nat/ra_nat.c
+++ linux-3.10.14/net/nat/hw_nat/ra_nat.c
@@ -4157,6 +4157,16 @@ static void SetAclFwd(uint32_t Ebl)
 #endif
 #endif
 		LanInt = ra_dev_get_by_name("br0");
+/* discard above!!  */
+#if defined (CONFIG_SUPPORT_OPENWRT)
+#if defined (CONFIG_RALINK_MT7621) || defined (CONFIG_ARCH_MT7623)
+		WanInt = ra_dev_get_by_name("eth1");
+#else
+		WanInt = ra_dev_get_by_name("eth0.2");
+#endif
+		LanInt = ra_dev_get_by_name("br-lan");
+#endif
+
 		for(i = 0;i< 6;i++){
 			value = RegRead(RALINK_ETH_SW_BASE + 0x2004 + (i * 0x100));
 			value |= (0x1 << 10);
